modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxSavepointsSelfTest.java

```diff
@@ -0,0 +1,290 @@
+
+                    tx.rollbackToSavepoint("s2");
+
+                    assertEquals("Failed in " + concurrency + ' ' + isolation + " transaction.",
+                        (Integer) 2, jcache().get("key"));
+
+                    tx.commit();
+                }
+
+                assertEquals("Failed in " + concurrency + ' ' + isolation + " transaction.",
+                    (Integer) 2, jcache().get("key"));
+            }
+    }
+
+    /**
+     * Tests valid and invalid rollbacks to savepoint.
+     */
+    public void testFailRollbackToSavepoint() {
+        for (final TransactionConcurrency concurrency : TransactionConcurrency.values())
+            for (final TransactionIsolation isolation : TransactionIsolation.values()) {
+                GridTestUtils.assertThrows(
+                    log, new Callable<Boolean>() {
+                        @Override public Boolean call() throws Exception {
+                            jcache().put("key", 0);
+
+                            try (Transaction tx = grid(0).transactions().txStart(concurrency, isolation)) {
+                                jcache().put("key", 1);
+
+                                tx.savepoint("s1");
+
+                                jcache().put("key", 2);
+
+                                tx.savepoint("s2");
+
+                                jcache().put("key", 3);
+
+                                tx.savepoint("s3");
+
+                                tx.rollbackToSavepoint("s2");
+
+                                assertEquals("Failed in " + concurrency + ' ' + isolation + " transaction.",
+                                    (Integer)2, jcache().get("key"));
+
+                                tx.rollbackToSavepoint("s3");
+                            }
+
+                            return null;
+                        }
+                    },
+                    IllegalArgumentException.class,
+                    "No such savepoint.");
+            }
+    }
+
+    /**
+     * Tests savepoint deleting.
+     */
+    public void testOverwriteSavepoints() {
+        for (final TransactionConcurrency concurrency : TransactionConcurrency.values())
+            for (final TransactionIsolation isolation : TransactionIsolation.values()) {
+                GridTestUtils.assertThrows(
+                    log, new Callable<Boolean>() {
+                        @Override public Boolean call() throws Exception {
+                            jcache().put("key", 0);
+
+                            try (Transaction tx = grid(0).transactions().txStart(concurrency, isolation)) {
+                                jcache().put("key", 1);
+
+                                tx.savepoint("s1");
+
+                                jcache().put("key", 2);
+
+                                tx.savepoint("s2");
+
+                                jcache().put("key", 3);
+
+                                tx.savepoint("s3");
+
+                                tx.savepoint("s1", true);
```

Comment:
may be write to javadoc in Transaction#savepoint(java.lang.String, boolean)
that , if u rewrite savepoint then all subsequent savepoints got erased.